core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -1442,56 +1442,93 @@ public boolean removePartition(long timestamp) {
-            LOG.error().$("partition is already detached [path=").$(path).I$();
-            return false;
-        }
+        // commit changes, there may be uncommitted rows of any partition
+        commit();
 
-        // when we want to delete first partition we must find out
-        // minTimestamp from next partition if it exists or next partition and so on
-        //
-        // when somebody removed data directories manually and then
-        // attempts to tidy up metadata with logical partition delete
-        // we have to uphold the effort and re-compute table size and its minTimestamp from
-        // what remains on disk
+        if (timestamp == getPartitionLo(maxTimestamp)) {
 
-        // find out if we are removing min partition
-        long nextMinTimestamp = minTimestamp;
-        if (timestamp == txWriter.getPartitionTimestamp(0)) {
-            nextMinTimestamp = readMinTimestamp(txWriter.getPartitionTimestamp(1));
-        }
-        long partitionNameTxn = txWriter.getPartitionNameTxnByPartitionTimestamp(timestamp);
-        columnVersionWriter.removePartition(timestamp);
+            // removing active partition
 
-        txWriter.beginPartitionSizeUpdate();
-        txWriter.removeAttachedPartitions(timestamp);
-        txWriter.setMinTimestamp(nextMinTimestamp);
-        txWriter.finishPartitionSizeUpdate(nextMinTimestamp, txWriter.getMaxTimestamp());
-        txWriter.bumpTruncateVersion();
+            if (index == 0) {
+                // removing the very last partition is equivalent to truncating the table
+                truncate();
+                return true;
+            }
 
-        columnVersionWriter.commit();
-        txWriter.setColumnVersion(columnVersionWriter.getVersion());
-        txWriter.commit(defaultCommitMode, denseSymbolMapWriters);
+            // calculate new transient row count, new min and max timestamp
+            final int prevIndex = index - 1;
+            final long prevTimestamp = txWriter.getPartitionTimestamp(prevIndex);
+            final long newTransientRowCount = txWriter.getPartitionSize(prevIndex);
+            try {
+                setPathForPartition(path.trimTo(rootLen), partitionBy, prevTimestamp, false);
+                TableUtils.txnPartitionConditionally(path, txWriter.getPartitionNameTxn(prevIndex));
+                readPartitionMinMax(ff, prevTimestamp, path, metadata.getColumnName(metadata.getTimestampIndex()), newTransientRowCount);
+            }
+            finally {
+                path.trimTo(rootLen);
+            }
 
-        // Call O3 methods to remove check TxnScoreboard and remove partition directly
-        safeDeletePartitionDir(timestamp, partitionNameTxn);
+            columnVersionWriter.removePartition(timestamp);
+            txWriter.beginPartitionSizeUpdate();
+            txWriter.removeAttachedPartitions(timestamp);
+            txWriter.setMaxTimestamp(attachMaxTimestamp);
```

Comment:
I think you don't need to call this and you can remove this new method `setMaxTimestamp`.
Next line sets max timestamp anyway. The only branch in the method you added

```
    public void setMinTimestamp(long timestamp) {
        recordStructureVersion++;
        minTimestamp = timestamp;
        if (prevMinTimestamp == Long.MAX_VALUE) {
            prevMinTimestamp = minTimestamp;
        }
    }
```

can happen when there are no partitions e.g. truncated table. This can never happen on drop partition, so the branch is unreachable.

Reply:
yes